File: 백준/Gold/17609. 회문/회문.py
import sys
input = sys.stdin.readline

def check_palindrome(text):
    left, right = 0, len(text)-1
    while left < right:
        if text[left] != text[right]:
            return False
        left += 1
        right -= 1
    return True

def check_pseudo_palindrome(text):
    is_left, is_right = True, True

    left, right = 0, len(text)-1
    flag = True
    while left < right:
        if text[left] != text[right]:
            if flag:
                left += 1
                flag = False
            else:
                is_left = False
                break
        else:
            left += 1
            right -= 1

    left, right = 0, len(text)-1
    flag = True
    while left < right:
        if text[left] != text[right]:
            if flag:
                right -= 1
                flag = False
            else:
                is_right = False
                break
        else:
            left += 1
            right -= 1
    
    return is_left or is_right

n = int(input())
for _ in range(n):
    str_input = input().strip()
    if check_palindrome(str_input):
        print(0)
    elif check_pseudo_palindrome(str_input):
        print(1)
    else:
        print(2)

File: 백준/Gold/17609. 회문/회문.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {

    public static boolean checkPalindrome(String text) {
        int left = 0;
        int right = text.length() - 1;

        while (left < right) {
            if (text.charAt(left) != text.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }

        return true;
    }

    public static boolean checkPseudoPalindrome(String text) {
        boolean isLeft = true;
        boolean isRight = true;

        int left = 0;
        int right = text.length() - 1;
        boolean flag = true;

        while (left < right) {
            if (text.charAt(left) != text.charAt(right)) {
                if (flag) {
                    left++;
                    flag = false;
                } else {
                    isLeft = false;
                    break;
                }
            } else {
                left++;
                right--;
            }
        }

        left = 0;
        right = text.length() - 1;
        flag = true;

        while (left < right) {
            if (text.charAt(left) != text.charAt(right)) {
                if (flag) {
                    right--;
                    flag = false;
                } else {
                    isRight = false;
                    break;
                }
            } else {
                left++;
                right--;
            }
        }

        return isLeft || isRight;
    }

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            String strInput = br.readLine().trim();

            if (checkPalindrome(strInput)) {
                System.out.println(0);
            } else if (checkPseudoPalindrome(strInput)) {
                System.out.println(1);
            } else {
                System.out.println(2);
            }
        }
    }
}